rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp

```diff
@@ -276,6 +276,68 @@ TYPED_TEST(TestAddCallbackGroupsToExecutor, one_node_many_callback_groups_many_e
   ASSERT_EQ(timer_executor.get_all_callback_groups().size(), 2u);
 }
 
+/*
+ * Test callback groups from one node to many executors.
+ * A subscriber on a new executor with a callback group not received a message
+ * because the executor can't be triggered while a subscriber created, see
+ * https://github.com/ros2/rclcpp/issues/1611
+*/
+TYPED_TEST(TestAddCallbackGroupsToExecutor, subscriber_triggered_to_receive_message)
+{
+  auto node = std::make_shared<rclcpp::Node>("my_node", "/ns");
+
+  // create a thread running an executor with a new callback group for a coming subscriber
+  rclcpp::CallbackGroup::SharedPtr cb_grp = node->create_callback_group(
+    rclcpp::CallbackGroupType::MutuallyExclusive, false);
+  rclcpp::executors::SingleThreadedExecutor cb_grp_executor;
+  std::thread cb_grp_thread = std::thread(
+    [&cb_grp, &node, &cb_grp_executor]() {
+      cb_grp_executor.add_callback_group(cb_grp, node->get_node_base_interface());
+      cb_grp_executor.spin();
+    });
+
+  // expect the subscriber to receive a message
+  bool received_message = false;
+  auto sub_callback = [&received_message](const test_msgs::msg::Empty::SharedPtr) {
+      received_message = true;
+    };
+
+  // a timer callback run on the default executor
+  rclcpp::TimerBase::SharedPtr timer = nullptr;
+  auto timer_callback = [&timer, &cb_grp, &node, &sub_callback]() {
+      if (timer) {
+        timer.reset();
+      }
+
+      // create a subscription using the `cb_grp` callback group
+      rclcpp::QoS qos = rclcpp::QoS(1).reliable();
+      auto options = rclcpp::SubscriptionOptions();
+      options.callback_group = cb_grp;
+      auto subscription =
+        node->create_subscription<test_msgs::msg::Empty>("topic_name", qos, sub_callback, options);
+
+      // create a publisher to send data
+      auto publisher =
+        node->create_publisher<test_msgs::msg::Empty>("topic_name", qos);
+      publisher->publish(test_msgs::msg::Empty());
+      // wait for message to be published
+      std::this_thread::sleep_for(100ms);
+    };
+
+  // timer on default callback group
+  timer = node->create_wall_timer(1s, timer_callback);
+  auto start = std::chrono::steady_clock::now();
+  while (!received_message &&
+    (std::chrono::steady_clock::now() - start) < 3s)
+  {
+    rclcpp::spin_some(node);
```

Comment:
I don't think it's important for the test, but this method will add and remove the node from an executor each time this is called, so it would probably be better to create an executor, add the node, and either spin in a thread or use spin until future complete with a timeout, rather than using one of the `rclcpp::spin*` functions.

Reply:
This kind of test code already existed in the `rclcpp` test, is it a problem here?

Reply:
No, but it's not ideal either. The tests are pretty scattered in terms of age and quality, so they're not necessarily the best example for best practices.

Like I said, it's probably not affecting the test, but using a future and `spin_until_future_complete` is better. Also, since this test is to do with how nodes, callback groups, and executors interact, it's probably better to not add/remove the node to the executor in a loop...